Support explicit comparison functions for iteration grouping

cquike opened the pull request:
A new constructor is created to support giving explicit comparison functions to the table sorting algorithms. This makes MSIter more generic. Eventually the other constructors could be removed if   all the code that uses MSIter (like VisibilityIterator* in CASA) moves to the new API.
Note that this new constructor doesn't support cached SORTED_TABLE, which could be considered legacy by now.

This work is in the context of CAS-8352 to allow the combination of SPW within subchunks.

cquike commented:
Note that this branch is based on PR #991, so it will be better if this gets merged after that other PR.
I will ask George Moellenbrock for a code review.

tammojan commented:
Sorry for only responding now! Now that the master is passing CI tests, could you rebase on the current master?

Also, I don't see any tests added. Is the new functionality tested?

cquike commented:
Hi,
I have added unit tests that exercise the new functionality. 
Additionally I have rebased on top of the latest #991 commit.
George Moellenbrock has reviewed the code and will soon post his comments.

gmoellen commented:
This PR continues from #991, and I endorse this for merging (after #991), modulo resolving the "Travis CI" built failure noted above.

-George

tammojan commented:
The travis failure is this time a real failure, the test `tMSIter` fails as follows:
```
    Start 379: tMSIter

379: Test command: /home/dijkema/opt/casacore/casacore/cmake/cmake_assay "./tMSIter"
379: Test timeout computed to be: 1500
379: 97,101d96
379: < ########
379: < nrow=30
379: < ########
379: < nrow=25
379: < nrow=5
379: FAIL (output not verified): ./tMSIter
1/1 Test #379: tMSIter ..........................***Failed    0.49 sec
```

cquike commented:
I have rebased this branch on top of the branch of #991 (which has been closed due to containing the same commits as this one). So now this branch contains two commits, one for #991 (https://github.com/cquike/casacore/commit/6192581b9a261386cb750dc195ca987339e86cae) and one for this PR itself (https://github.com/cquike/casacore/commit/2e0a730d2d0f811dc436afc1fac74a4a53e5ca68).
I have fixed the unit tests. It was simply that I forgot to add the new stdout of the unit test that I have added.
So I think that now the branch is really ready to be merged.

tammojan commented:
Thanks for all the work @cquike, sorry this took so long to get merged!

cquike commented:
Thank you and thanks also @gmoellen  for the code reviews!!
